local actionRow = require("@self/message/components/actionRow")
local activity = require("@self/activity")
local allowedMention = require("@self/message/allowedMention")
local answer = require("@self/message/poll/answer")
local attachment = require("@self/attachment")
local author = require("@self/embed/author")
local automoderationAction = require("@self/automoderation/action/action")
local automoderationActionMetadata = require("@self/automoderation/action/metadata")
local automoderationRule = require("@self/automoderation/rule")
local automoderationTriggerMetadata = require("@self/automoderation/trigger/metadata")
local button = require("@self/message/components/button")
local channel = require("@self/channel")
local choice = require("@self/interaction/choice")
local defaultReaction = require("@self/default/reaction")
local defaultValue = require("@self/message/components/selectMenu/defaultValue")
local embed = require("@self/embed/embed")
local emoji = require("@self/emoji")
local field = require("@self/embed/field")
local footer = require("@self/embed/footer")
local forumMedia = require("@self/thread/forumMedia")
local forumTag = require("@self/forumTag")
local guild = require("@self/guild/guild")
local image = require("@self/embed/image")
local intents = require("@self/intents")
local interaction = require("@self/interaction/interaction")
local invite = require("@self/invite")
local media = require("@self/message/poll/media")
local message = require("@self/message/message")
local messageParams = require("@self/thread/messageParams")
local messageThread = require("@self/thread/messageThread")
local modal = require("@self/message/components/modal")
local noMessageThread = require("@self/thread/noMessageThread")
local onboarding = require("@self/guild/onboarding/onboarding")
local onboardingOption = require("@self/guild/onboarding/option")
local option = require("@self/interaction/option")
local overwrite = require("@self/guild/overwrite")
local permission = require("@self/permission")
local poll = require("@self/message/poll/poll")
local presence = require("@self/presence")
local prompt = require("@self/guild/onboarding/prompt")
local provider = require("@self/embed/provider")
local reference = require("@self/message/reference")
local role = require("@self/guild/role")
local selectMenu = require("@self/message/components/selectMenu/selectMenu")
local selectMenuOption = require("@self/message/components/selectMenu/option")
local textInput = require("@self/message/components/textInput")
local thumbnail = require("@self/embed/thumbnail")
local video = require("@self/embed/video")
local welcomeScreen = require("@self/guild/welcomeScreen/welcomeScreen")
local welcomeScreenChannel = require("@self/guild/welcomeScreen/channel")
local widgetSettings = require("@self/guild/widgetSettings")

export type providerJSON = provider.JSON
export type fieldJSON = field.JSON
export type thumbnailJSON = thumbnail.JSON
export type embedJSON = embed.JSON
export type authorJSON = author.JSON
export type imageJSON = image.JSON
export type footerJSON = footer.JSON
export type videoJSON = video.JSON
export type inviteJSON = invite.JSON
export type emojiJSON = emoji.JSON
export type channelJSON = channel.JSON
export type messageParamsJSON = messageParams.JSON
export type messageThreadJSON = messageThread.JSON
export type forumMediaJSON = forumMedia.JSON
export type noMessageThreadJSON = noMessageThread.JSON
export type guildJSON = guild.JSON
export type onboardingOptionJSON = onboardingOption.JSON
export type promptJSON = prompt.JSON
export type onboardingJSON = onboarding.JSON
export type welcomeScreenChannelJSON = welcomeScreenChannel.JSON
export type welcomeScreenJSON = welcomeScreen.JSON
export type overwriteJSON = overwrite.JSON
export type widgetSettingsJSON = widgetSettings.JSON
export type roleJSON = role.JSON
export type forumTagJSON = forumTag.JSON
export type actionRowJSON = actionRow.JSON
export type textInputJSON = textInput.JSON
export type buttonJSON = button.JSON
export type selectMenuOptionJSON = selectMenuOption.JSON
export type defaultValueJSON = defaultValue.JSON
export type selectMenuJSON = selectMenu.JSON
export type mediaJSON = media.JSON
export type answerJSON = answer.JSON
export type pollJSON = poll.JSON
export type messageJSON = message.JSON
export type referenceJSON = reference.JSON
export type allowedMentionJSON = allowedMention.JSON
export type optionJSON = option.JSON
export type choiceJSON = choice.JSON
export type interactionJSON = interaction.JSON
export type attachmentJSON = attachment.JSON
export type defaultReactionJSON = defaultReaction.JSON
export type permissionJSON = permission.JSON
export type activityJSON = activity.JSON
export type presenceJSON = presence.JSON
export type AutomoderationRuleJSON = automoderationRule.JSON
export type AutomoderationActionJSON = automoderationAction.JSON
export type AutomoderationActionMetadataJSON = automoderationActionMetadata.JSON
export type AutomoderationTriggerMetadataJSON = automoderationTriggerMetadata.JSON
export type ModalJSON = modal.JSON
export type intents = intents.Intents

return {
    intents = intents,
    permission = permission,
    defaultReaction = defaultReaction,
    attachment = attachment,
    presence = presence,
    activity = activity,
    automoderation = {
        rule = automoderationRule,
        action = {
            action = automoderationAction,
            metadata = automoderationActionMetadata,
        },
        triggerMetadata = automoderationTriggerMetadata,
    },
    interaction = {
        interaction = interaction,
        choice = choice,
        option = option,
    },
    message = {
        allowedMention = allowedMention,
        reference = reference,
        message = message,
        poll = {
            poll = poll,
            answer = answer,
            media = media,
        },
        components = {
            selectMenu = {
                selectMenu = selectMenu,
                defaultValue = defaultValue,
                option = selectMenuOption,
            },
            modal = modal,
            button = button,
            textInput = textInput,
            actionRow = actionRow,
        },
    },
    forumTag = forumTag,
    guild = {
        role = role,
        widgetSettings = widgetSettings,
        overwrite = overwrite,
        welcomeScreen = {
            welcomeScreen = welcomeScreen,
            channel = welcomeScreenChannel,
        },
        onboarding = {
            onboarding = onboarding,
            prompt = prompt,
            option = onboardingOption,
        },
        guild = guild,
    },
    thread = {
        noMessageThread = noMessageThread,
        forumMedia = forumMedia,
        messageThread = messageThread,
        messageParams = messageParams,
    },
    channel = channel,
    emoji = emoji,
    invite = invite,
    embed = {
        video = video,
        footer = footer,
        image = image,
        author = author,
        embed = embed,
        thumbnail = thumbnail,
        field = field,
        provider = provider,
    },
}
